### PR TITLE
[CR 151] port edgeR functions necessary for differential expression

### PR DESCRIPTION
**Pull request by EpigeneMax**
## Description
Part of the edgeR package was already ported to support `pycombat_seq`.
We now wish to port enough of edgeR to be able to use its differential expression features.

## Motivation and Context
Nicely complements the previous work on porting diff exp from DESeq2.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] No code (non-breaking change that does not impact code: formatting, build system, documentation...)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] My changes are in a separate branch named after a Jira card, or -- if not applicable -- with a descriptive name.
- [x] I have descriptive commit messages, with a short title (first line).
- [x] I have added and/or updated tests to reflect my changes.
- [x] I have reviewed and updated the documentation accordingly.
- [x] I have updated the [Changelog](../CHANGELOG.md).
- [x] My code and documentation follow the [numpy docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
- [x] I have reviewed possible licensing issues, preferably with a person in charge of the topic.

NB: we also port some code from limma, which is released under GPL>=2, which is compatible with `inmoose` GPL3 license.

### CODE REVIEW

File: inmoose/edgepy/edgepy_cpp/__init__.pxd [hunk 1]
```diff
@@ -22,6 +22,7 @@ from numpy cimport ndarray
 # cf. file src/utils.h from edgeR source repo
 
 cdef public ndarray[double, ndim=1] vector2ndarray "vector2ndarray"(const vector.vector[double]& data)
+cdef public double compute_unit_nb_deviance(double, double, double)
 
 cdef extern from "objects.h":
     cpdef bool is_integer_array(ndarray arr) except +
```
[the remark this answers is not in the document]
Comment (soleneweill) [reply inmoose/edgepy/edgepy_cpp/__init__.pxd:28]: And also can you briefly explain the difference between cpp and pyx file?

File: inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx [hunk 2]
```diff
@@ -77,3 +88,80 @@ cdef public void f77_dpotrs "f77_dpotrs"(const char *uplo, const int *n, const i
...
+    y = y + mildly_low_value
+    mu = mu + mildly_low_value
+
+    # Calculating the deviance using either the Poisson (small phi*mu), the
+    # Gamma (large phi*mu) or NB (everything else). Some additional work is
+    # put in to make the transitiosn between families smooth
+    resid = y - mu
+    product = mu * phi
+
+    if phi < 1e-4:
```
Comment (soleneweill) [inline inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx:131]: How can an array be smaller than a single value ?

File: inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx [hunk 3]
```diff
@@ -77,3 +88,80 @@ cdef public void f77_dpotrs "f77_dpotrs"(const char *uplo, const int *n, const i
...
+        return 2 * (y * log(y/mu) - resid - 0.5*resid*resid*phi*(1+phi*(2/3*resid-y)))
+    elif product > 1e6:
+        return 2 * (resid/mu - log(y/mu)) * mu/(1+product)
+    else:
+        return 2 * (y * log(y/mu) + (y + 1/phi) * log((mu + 1/phi)/(y + 1/phi)))
+
+@cython.ufunc
+cdef double _q2qnbinom(double x, double input_mean, double output_mean, double dispersion):
```
Comment (soleneweill) [inline inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx:139]: documentation ?

File: inmoose/edgepy/exactTest.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+            dispersion = self.trended_dispersion
+        elif dispersion == "tagwise":
+            dispersion = self.tagwise_dispersion
+        elif dispersion == "auto":
+            dispersion = self.getDispersion()
+        else:
+            raise ValueError(f"Invalid value {dispersion} for 'dispersion'")
+
+        if dispersion is None:
+            raise ValueError("specified dispersion not found in object")
```
Comment (soleneweill) [inline inmoose/edgepy/exactTest.py:162]: How could that happened as you already have a condition to test if the dispersion is None and input a value if so?

File: inmoose/edgepy/exactTestBetaApprox.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        matrix of counts for the first of the two experimental groups to be
+        tested for differences. Rows correspond to genes and columns to
+        libraries. Libraries are assumed to be equal in size -- *e.g.* adjusted
+        pseudocounts from the output of :func:`equalizeLibSizes`.
+    y2 : matrix
+        matrix of counts for the second of the two experimental groups to be
+        tested for differences. Rows correspond to genes and columns to
+        libraries. Libraries are assumed to be equal in size -- *e.g.* adjusted
+        pseudocounts from the output of :func:`equalizeLibSizes`.
+    dispersion : array_like of floats
```
Comment (soleneweill) [inline inmoose/edgepy/exactTestBetaApprox.py:48]: Here you have array_like of floats as type but the default value is int. Is it normal?

File: inmoose/edgepy/exactTestByDeviance.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        matrix of counts for the first of the two experimental groups to be
+        tested for differences. Rows correspond to genes and columns to
+        libraries. Libraries are assumed to be equal in size -- *e.g.* adjusted
+        pseudocounts from the output of :func:`equalizeLibSizes`.
+    y2 : matrix
+        matrix of counts for the second of the two experimental groups to be
+        tested for differences. Rows correspond to genes and columns to
+        libraries. Libraries are assumed to be equal in size -- *e.g.* adjusted
+        pseudocounts from the output of :func:`equalizeLibSizes`.
+    dispersion : array_like of floats
```
Comment (soleneweill) [inline inmoose/edgepy/exactTestByDeviance.py:53]: Same here

File: inmoose/edgepy/residDF.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        if k not in d:
+            d[k] = []
+        d[k].append(i)
+
+    return [v for v in d.values()]
+
+
+def _residDF(zero, design):
+    """
+    Effective residual degrees of freedom after adjusting for exact zeros
```
Comment (soleneweill) [inline inmoose/edgepy/residDF.py:41]: incomplete function documentation

File: inmoose/edgepy/residDF.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+# -----------------------------------------------------------------------------
+
+# This file is based on the file 'R/residDF.R' of the Bioconductor edgeR package (version 3.38.4).
+
+import numpy as np
+
+
+def _comboGroups(truths):
+    """
+    Function that returns a list of vectors of indices, where each vector refers to the rows with the same combination of True/False values in :code:`truths`.
```
Comment (soleneweill) [inline inmoose/edgepy/residDF.py:26]: incomplete documentation

File: inmoose/limma/fitFDist.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            s20 = x.mean()
+        else:
+            s20 = emean.exp()
+
+    return {"scale": s20, "df2": df2}
+
+
+def trigammaInverse(x):
+    """
+    Solve trigamma(y) = x for y
```
Comment (soleneweill) [inline inmoose/limma/fitFDist.py:203]: incomplete documentation

File: inmoose/limma/squeezeVar.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    var_post = _squeezeVar(var=var, df=df, var_prior=fit["scale"], df_prior=df_prior)
+
+    return {"var_post": var_post, "var_prior": fit["scale"], "df_prior": df_prior}
+
+
+def _squeezeVar(var, df, var_prior, df_prior):
+    """
+    Squeeze posterior variances given hyperparameters
+
+    NAs not allowed in :code:`df_prior`
```
Comment (soleneweill) [inline inmoose/limma/squeezeVar.py:118]: incomplete documentation

File: inmoose/edgepy/aveLogCPM.py [hunk 11]
```diff
@@ -84,6 +84,20 @@ def aveLogCPM_DGEList(self, normalized_lib_sizes=True, prior_count=2, dispersion
     )
 
 
+def aveLogCPM_DGEGLM(y, prior_count=2, dispersion=None):
```
[the remark this answers is not in the document]
Comment (EpigeneMax) [reply inmoose/edgepy/aveLogCPM.py:87]: 👍

File: inmoose/edgepy/edgepy_cpp/__init__.pxd (hunk 1 above)
[the remark this answers is not in the document]
Comment (EpigeneMax) [reply inmoose/edgepy/edgepy_cpp/__init__.pxd:28]: `.pxd` and `.pyx` files are Cython files, from which C/C++ code is generated.
`.pxd` files contain declarations, and are used to generate the C/C++ header files (`.h` files). Headers make sense for statically typed compiled languages (such as C/C++), but are not relevant for dynamically typed interpreted languages such as Python. `.pyi` are Python headers but are very rarely used.
`.pyd` files contain implementations, in a language close to Python with annotations to guide C/C++ code generation. They are used to generate `.c` (C) or `.cpp` (C++) files.

Note that Cython also generate Python bindings for the generated C/C++ code, a feature we use here: `compute_unit_nb_deviance` is needed in C++ code and in Python code.

File: inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx (hunk 2 above)
Comment (EpigeneMax) [reply inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx:131]: This function is a UFunc (see the `cython.ufunc` decorator): its code describe what to do on individual values, and Cython takes care of the vectorization to call it on arrays.
So `phi` here is a `double` (as per the function signature), although the docstring says it is an array.

File: inmoose/edgepy/exactTest.py (hunk 4 above)
Comment (EpigeneMax) [reply inmoose/edgepy/exactTest.py:162]: Double check the `if` nesting structure.
If `dispersion` is `None`, then it is assigned the value `"auto"` (ligne 147).
Then, `str` values are dispatched, but there is no guarantee that the dispatch will fetch a non-`None` value. Typically if the dispersion have never been computed in `self`.

File: inmoose/edgepy/exactTestBetaApprox.py (hunk 5 above)
Comment (EpigeneMax) [reply inmoose/edgepy/exactTestBetaApprox.py:48]: Not a big deal, as the `int`s will be converted when combined with the other `float` values.
Fixed the default value to `0.` though.

File: inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx (hunk 3 above)
Comment (EpigeneMax) [reply inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx:139]: done

**Comment by EpigeneMax** [comment]
Be my guest.

**Pull request merged**